### PR TITLE
Reduce complexity of cypress commands

### PR DESCRIPTION
**Pull request by max-nextcloud**
### 📝 Summary

With time we build up quite a bit of complexity in cypress commands that is not necessary.
This PR tries to reduce it step by step, commit by commit.

Best reviewed on a per commit basis I think.

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [ ] Tests (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [ ] Documentation ([README](https://github.com/nextcloud/collectives/blob/main/README.md) or [documentation](https://github.com/nextcloud/collectives/blob/main/docs/)) has been updated or is not required

### CODE REVIEW

File: cypress/support/commands.js [hunk 1]
```diff
@@ -130,6 +130,20 @@ Cypress.Commands.add('dispatch', (...args) => {
 		.invoke(silent, ...['dispatch', ...args])
 })
 
+Cypress.Commands.add('findBy',
```
Comment (mejo-) [inline cypress/support/commands.js:157]: Hehe, well I don't know exactly if this "reduce[s] complexity of cypress commands" :sweat_smile:

I find this code pretty hard to parse.

**Review by mejo- — APPROVED** [review]
Wow, thanks for diving so deep into improving the logic of our custom Cypress commands. Code changes look great and quite sophisticated (to an extend that I struggled a bit to follow the logic).

File: cypress/support/commands.js [hunk 2]
```diff
@@ -169,52 +186,43 @@ Cypress.Commands.add('deleteCollective', (name) => {
...
+	cy.store('state.collectives.collectives')
+		.invoke('find', c => c.name === name)
+		.then(({ id }) => cy.dispatch(UPDATE_COLLECTIVE_PAGE_MODE, { id, mode }))
 })
 
 /**
  * Add a page to a collective
  */
-Cypress.Commands.add('seedPage', (name, parentFilePath, parentFileName) => {
+Cypress.Commands.add('seedPage', function(name, parentFilePath, parentFileName) {
```
[the remark this answers is not in the document]
Comment (max-nextcloud) [reply cypress/support/commands.js:245]: Good catch. I was experimenting with using `this.something` to access an alias. So that's probably a leftover from that. I'll revert it.

File: cypress/support/commands.js [hunk 3]
```diff
@@ -103,21 +104,30 @@ Cypress.Commands.add('enableDashboardWidget', (widgetName) => {
 	)
 })
 
-Cypress.Commands.add('store', (selector) => {
+Cypress.Commands.add('store', (selector, options = {}) => {
 	if (selector) {
-		cy.window().its(`app.$store.${selector}`)
+		Cypress.log()
+	}
+	if (selector) {
```
[the remark this answers is not in the document]
Comment (max-nextcloud) [reply cypress/support/commands.js:118]: Oh.. I thought i had dropped the former one in a later commit. But maybe i reverted that. I'll take a look.

[the remark this answers is not in the document]
Comment (mejo-) [reply cypress/support/commands.js:118]: You're right, it's reverted in a later commit. So nevermind.

File: cypress/support/commands.js [hunk 4]
```diff
@@ -222,22 +222,15 @@ Cypress.Commands.add('seedPage', (name, parentFilePath, parentFileName) => {
  * Upload a file
  */
 Cypress.Commands.add('uploadFile', (path, mimeType, remotePath = '') => {
+	Cypress.log()
```
[the remark this answers is not in the document]
Comment (max-nextcloud) [reply cypress/support/commands.js:266]: It logs the name of the command with the arguments given just like cypress logs its own commands:

|before|after|
|--------|-------|
|grafik|grafik|

File: cypress/support/commands.js (hunk 1 above)
Comment (max-nextcloud) [reply cypress/support/commands.js:157]: Fair enough. :)
I'm happy to explain it if you're interested.

Do you also find the usage hard to parse?
```js
Cypress.Commands.add('deleteCollective', (name) => {
    cy.dispatch(GET_COLLECTIVES)
    cy.store('state.collectives.collectives')
        .findBy({ name })
        .dispatch(TRASH_COLLECTIVE)
...
```
I understand that all three of these (`store`, `findBy` and `dispatch`) are very abstract and generic. I also find them quite powerful.

I think some of the complexity and also power comes from the use of *"child actions"*. They take the output of the other action as input - but that's pretty implicit so harder to parse. But I think i got used to it very fast.

The last line could also be written as...
```js
    .then(collective => {
      cy.dispatch(TRASH_COLLECTIVE, collective)
    })
```

My impression was that one ends up with lot's of `.then`s, `cy.`s and a deep nesting pretty quickly.

Comment (max-nextcloud) [reply cypress/support/commands.js:157]: I'll merge this for now and continue exploring alternatives.

Comment (mejo-) [reply cypress/support/commands.js:157]: Absolutely fine with merging, thanks! And the usage is super straight-forward. I'm not opposed to those abstract functions at all, just wanted to mention that they bring quite some complexity. Thanks a lot for the extra explanation. After going through them a few times I think I understood their logic :innocent:
